src/write_statsd.c

```diff
@@ -0,0 +1,476 @@
+  char*  event_line_format;
+  size_t event_line_base_len;
+  /* The short line format is used when the DS name is not included. */
+  char*  short_event_line_format;
+  size_t short_event_line_base_len;
+};
+typedef struct write_statsd_config_s write_statsd_config_t;
+
+struct write_statsd_link_s {
+  struct sockaddr_in server;
+  int    sock;
+};
+typedef struct write_statsd_link_s write_statsd_link_t;
+
+
+/*** Module variables. ***/
+static char* DS_TYPE_TO_STATSD[] = {
+  "c",  /* DS_TYPE_COUNTER  */
+  "g",  /* DS_TYPE_GAUGE    */
+  "g",  /* DS_TYPE_DERIVE   */
+  NULL  /* DS_TYPE_ABSOLUTE */
+};
+
+/*
+ * The configuration is stored in a global instance and must not
+ * be modified outside of load_config or initialise.
+ */
+static write_statsd_config_t configuration = {
+  NULL, 8125, NULL, NULL, 0, 0,
+  NULL, 0, NULL, 0
+};
+
+
+/*** Module functions. ***/
+static void* allocate(size_t size) {
+  void* mem = (char*)malloc(size);
+  if (mem == NULL) {
+    ERROR("write_statsd plugin: not enough memory.");
+    return NULL;
+  }
+  return mem;
+}
+
+static char* ds_value_to_string(int type, value_t value) {
+  char* result = allocate(VALUE_STR_LEN);
+  if (result == NULL) {
+    return NULL;
+  }
+
+  switch(type) {
+    case DS_TYPE_COUNTER:
+      ssnprintf(result, VALUE_STR_LEN, "%lld", value.counter);
+      break;
+
+    case DS_TYPE_GAUGE:
+      ssnprintf(result, VALUE_STR_LEN, WS_DOUBLE_FORMAT, value.gauge);
+      break;
+
+    case DS_TYPE_DERIVE:
+      ssnprintf(result, VALUE_STR_LEN, "%"PRId64, value.derive);
+      break;
+
+    default:
+      free(result);
+      return NULL;
+  }
+
+  return result;
+}
+
+
+static write_statsd_link_t open_socket(void) {
+  write_statsd_link_t link  = {{0}, 0};
+  struct addrinfo* resolved = NULL;
+  int error = 0;
+
+  link.sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+  if (link.sock == -1) {
+    ERROR("write_statsd plugin: unable to open socket connection.");
+    link.sock = 0;
+    return link;
+  }
+
+  link.server.sin_family = AF_INET;
+  link.server.sin_port   = htons(configuration.port);
+  error = getaddrinfo(configuration.host, NULL, NULL, &resolved);
+
+  if (error) {
+    ERROR("write_statsd plugin: unable to resolve address - %s.",
+          gai_strerror(error));
+    close(link.sock);
+    link.sock = 0;
+    return link;
+  }
+
+  memcpy(&link.server.sin_addr,
+         &((struct sockaddr_in*)resolved->ai_addr)->sin_addr,
+         sizeof(struct in_addr));
+  freeaddrinfo(resolved);
+  return link;
+}
+
+
+static int write_statsd_config(oconfig_item_t *conf) {
+  int idx;
+  int status;
+
+  for (idx = 0; idx < conf->children_num; idx++) {
+    oconfig_item_t *child = conf->children + idx;
+    status = 0;
+
+    if (strcasecmp("Host", child->key) == 0) {
+      status = cf_util_get_string(child, &configuration.host);
+
+    } else if (strcasecmp("Port", child->key) == 0) {
+      status = cf_util_get_int(child, & configuration.port);
```

Comment:
Please use `cf_util_get_service()`. The returned string should be passed to `getaddrinfo()` (see above).